connection.go

```diff
@@ -1666,3 +1735,52 @@ func (conn *Connection) ServerProtocolInfo() ProtocolInfo {
 func (conn *Connection) ClientProtocolInfo() ProtocolInfo {
 	return clientProtocolInfo.Clone()
 }
+
+func shutdownEventCallback(event WatchEvent) {
+	// Receives "true" on server shutdown.
+	// See https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/graceful_shutdown/
+	// step 2.
+	val, ok := event.Value.(bool)
+	if ok && val {
+		go event.Conn.processShutdown()
+	}
+}
+
+func (conn *Connection) processShutdown() {
+	// Forbid state changes.
+	conn.mutex.Lock()
+	defer conn.mutex.Unlock()
+
+	if !atomic.CompareAndSwapUint32(&(conn.state), connConnected, connShutdown) {
+		return
+	}
+	conn.notify(Shutdown)
+
+	c := conn.c
+	for (atomic.LoadUint32(&(conn.state)) == connShutdown) &&
+		(atomic.LoadInt64(&(conn.requestCnt)) != 0) &&
+		(c == conn.c) {
+		// Use cond var on conn.mutex since request execution may
+		// call reconnect(). It is ok if state changes as part of
+		// reconnect since Tarantool server won't allow to reconnect
+		// in the middle of shutting down.
+		conn.cond.Wait()
+	}
+	// Do not unregister task explicitly here since connection teardown
+	// has the same effect. To clean up connection resources,
+	// unregister on full close.
+
+	if (atomic.LoadUint32(&(conn.state)) == connShutdown) &&
+		(c == conn.c) {
+		// Start to reconnect based on common rules, same as in net.box.
+		// Reconnect also closes the connection: server waits until all
+		// subscribed connections are terminated.
+		// See https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/graceful_shutdown/
+		// step 3.
+		conn.reconnectImpl(
+			ClientError{
+				ErrConnectionClosed,
+				"connection closed after server shutdown",
+			}, conn.c)
+	}
+}
```

Comment:
It looks a little easier:
```suggestion
	for {
		if atomic.LoadUint32(&(conn.state)) != connShutdown || c != conn.c) {
			return
		}
		If atomic.LoadInt64(&(conn.requestCnt)) == 0 {
			break
		}
		// Use cond var on conn.mutex since request execution may
		// call reconnect(). It is ok if state changes as part of
		// reconnect since Tarantool server won't allow to reconnect
		// in the middle of shutting down.
		conn.cond.Wait()
	}

	// Start to reconnect based on common rules, same as in net.box.
	// Reconnect also closes the connection: server waits until all
	// subscribed connections are terminated.
	// See https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/graceful_shutdown/
	// step 3.
	conn.reconnectImpl(
		ClientError{
			ErrConnectionClosed,
			"connection closed after server shutdown",
		}, conn.c)
}
```

Reply:
Thank you for the good advice, reworked.